examples/wiegand/main.c

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright (C) 2019 Hackerspace San Salvador
+ *
+ * This example demonstrates the use of the wiegand driver to
+ * decode a signal generated within the same example.
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     examples
+ * @{
+ *
+ * @file
+ * @brief       Application to demonstrates the use of the Wiegand driver
+ *
+ * @author      Mario Gomez <mario.gomez@teubi.co>
+ *
+ * @}
+ */
+#include <stdio.h>
+#include <string.h>
+
+#include "thread.h"
+#include "xtimer.h"
+#include "periph/gpio.h"
+#include "shell.h"
+#include "wiegand.h"
+#include "wiegand_params.h"
+#include "msg.h"
+
+/* BEGIN: Global parameters */
+
+// To test connect WG_OUT_D1 with WG_IN_D1 and WG_OUT_D0 with WG_IN_D0
+
+/*
+ * Configuration for Arduino MKR1000 or Arduino Zero
+ * It should support any GPIO with flank-triggered interrupts.
+ *
+ */
+#define WG_IN_D0 GPIO_PIN(PA, 4) // Pin A3 in MRK1000 / Arduino Zero
+#define WG_IN_D1 GPIO_PIN(PA, 6) // Pin A5 in MRK1000 / Arduino Zero
+#define WG_OUT_D0 GPIO_PIN(PA, 5) // Pin A4 in MRK1000 / Arduino Zero
+#define WG_OUT_D1 GPIO_PIN(PA, 7) // Pin A6 in MRK1000 / Arduino Zero
+
+/* Configuration for SAMR21G18-based boards */
+//#define WG_IN_D0 GPIO_PIN(PA, 6)
+//#define WG_IN_D1 GPIO_PIN(PA, 5)
+//#define WG_OUT_D0 GPIO_PIN(PA, 8)
+//#define WG_OUT_D1 GPIO_PIN(PA, 7)
+
+/* END: Global parameters */
+
+/* BEGIN: Wiegand reader thread */
+char wg_reader_stack[THREAD_STACKSIZE_DEFAULT];
+void *wg_reader_thread(void *arg)
+{
+  (void) arg;
+  uint32_t card;
+
+  // Create a new device
+  wiegand_t wiegand_dev;
+
+  // Configure connection parameters
+  wiegand_params_t wiegand_param = {
+    .d0 = WG_IN_D0,
+    .d1 = WG_IN_D1,
+    .flank = GPIO_RISING
+  };
+
+  // Note: Default wiegand_params can be changed on the Makefile
+  wg_init(&wiegand_dev, &wiegand_param);
+
+  // Note: uncomment the following to use the defaul parameters
+  // requires to include the file "wiegand_params.h"
+  // defaults could be changed on the Makefile
+  //wg_init(&wiegand_dev, &wiegand_params);
+
+  while(1) {
+    if(wg_available(&wiegand_dev)) {
```

Comment:
Instead of polling, you could register a callback with your driver that gets called when data is available.

Reply:
Another option is to make `wg_get_code()` blocking - see my other comment.

Reply:
The indication for end-of-frame in wiegand is no-activity on any of the two data lines (D0 or D1) for more than 20ms. I was thinking originally in a callback but then I would need to start/reset a process (or a timer) that is activated *after* a level-change is detected on any of the lines.

I'm not clear for me if it's safe in RIOT to try to start/reset a timer within an interrupt context.

Reply:
Current implementations checks if there is a new code in the buffer each time you try to read it. Buffer is always updated when activity is detected on the line. I decided it would be better to leave the user decide how to manage the blocking instead implementing this into the driver. However I would like to know the feedback of other users to see what they think.